stream_win.c

```diff
@@ -46,79 +63,79 @@ typedef BOOL (WINAPI *get_overlapped_result_ex_func)(
     BOOL alertable);
 static get_overlapped_result_ex_func get_overlapped_result_ex;
 
-static BOOL WINAPI win7_get_overlapped_result_ex(
+static BOOL WINAPI get_overlapped_result_ex_impl(
     HANDLE file,
     LPOVERLAPPED olap,
     LPDWORD bytes,
     DWORD millis,
     BOOL alertable) {
 
-  while (true) {
-    if (GetOverlappedResult(file, olap, bytes, FALSE)) {
-      // Result is available
-      return TRUE;
-    }
+  DWORD waitReturnCode, err;
 
-    ULONGLONG start = GetTickCount64();
-    if (SleepEx(millis, alertable) == WAIT_IO_COMPLETION) {
+  stream_debug( "Preparing to wait for maximum %ums\n", millis );
+  if ( millis != 0 ) {
+
+    waitReturnCode = WaitForSingleObjectEx(olap->hEvent, millis, alertable);
+    switch (waitReturnCode)
+    {
+    case WAIT_OBJECT_0:
+      // Event is signaled, overlapped IO operation result should be available.
+      break;
+    case WAIT_IO_COMPLETION:
+      // WaitForSingleObjectEx returnes because the system added an I/O completion
+      // routine or an asynchronous procedure call (APC) to the thread queue.
       SetLastError(WAIT_IO_COMPLETION);
+      break;
+    case WAIT_TIMEOUT:
+      // We reached the maximum allowed wait time, the IO operation failed
+      // to complete in timely fashion.
+      SetLastError(WAIT_TIMEOUT);
+      return FALSE;
+
+    case WAIT_FAILED:
+      // something went wrong calling WaitForSingleObjectEx
+      err = GetLastError();
+      stream_debug("WaitForSingleObjectEx failed: %s\n", win32_strerror(err));
+      return FALSE;
+
+    default:
+      // unexpected situation deserving investigation.
+      err = GetLastError();
+      stream_debug("Unexpected error: %s\n", win32_strerror(err));
       return FALSE;
-    }
-    ULONGLONG end = GetTickCount64();
-
-    if (millis != INFINITE) {
-      millis = (DWORD)((ULONGLONG)millis - (end - start));
-      if (millis <= 0) {
-        // Out of time
-        SetLastError(WAIT_TIMEOUT);
-        return FALSE;
-      }
     }
   }
+
+  return GetOverlappedResult(file, olap, bytes, FALSE);
 }
 
+
 static BOOL WINAPI probe_get_overlapped_result_ex(
     HANDLE file,
     LPOVERLAPPED olap,
     LPDWORD bytes,
     DWORD millis,
-    BOOL alertable) {
+    BOOL alertable ) {
   get_overlapped_result_ex_func func;
 
   func = (get_overlapped_result_ex_func)GetProcAddress(
       GetModuleHandle("kernel32.dll"),
       "GetOverlappedResultEx");
 
-  if (!func) {
-    func = win7_get_overlapped_result_ex;
+  if ( getenv( "WATCHMAN_WIN7_COMPAT") || !func ) {
```

Comment:
would be great to explicitly check for the env var containing "1" so that we don't chose to do this if someone set it to 0.

Just adding ` && getenv("WATCHMAN_WIN7_COMPAT")[0] == '1'` in that conditional is probably good enough if you're feeling too lazy to add a local variable for it :-)